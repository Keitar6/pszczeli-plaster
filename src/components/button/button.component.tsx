import { ButtonHTMLAttributes, FC } from "react";
import {
  BaseButton,
  ButtonSpiner,
  GoogleSignInButton,
  InvertedButton,
  UserMenuFuncButton,
  LoginButton,
  InputBar,
  ProductCard,
  SortingTypes,
  CartFuncButton
} from "./button.styles";

export enum BUTTON_TYPE_CLASSES {
  base = "base",
  google = "google-sign-in",
  inverted = "inverted",
  userMenuFuncButton = "userMenuFuncButton",
  login = "LoginButton",
  InputBar = "InputBar",
  productCard = "ProductCard",
  cartFuncButton = "CartFuncButton",
  sorting = "Sorting"
}

type ButtonProps = {
  buttonType?: BUTTON_TYPE_CLASSES;
  isLoading?: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const getButton = (buttonType = BUTTON_TYPE_CLASSES.base): typeof BaseButton =>
  ({
    [BUTTON_TYPE_CLASSES.base]: BaseButton,
    [BUTTON_TYPE_CLASSES.google]: GoogleSignInButton,
    [BUTTON_TYPE_CLASSES.login]: LoginButton,
    [BUTTON_TYPE_CLASSES.inverted]: InvertedButton,
    [BUTTON_TYPE_CLASSES.userMenuFuncButton]: UserMenuFuncButton,
    [BUTTON_TYPE_CLASSES.InputBar]: InputBar,
    [BUTTON_TYPE_CLASSES.productCard]: ProductCard,
    [BUTTON_TYPE_CLASSES.sorting]: SortingTypes,
    [BUTTON_TYPE_CLASSES.cartFuncButton]: CartFuncButton
  }[buttonType]);

const Button: FC<ButtonProps> = ({
  children,
  buttonType,
  isLoading,
  ...otherProps
}) => {
  const CustomButton = getButton(buttonType);
  return (
    <CustomButton disabled={isLoading} {...otherProps}>
      {isLoading ? <ButtonSpiner /> : children}
    </CustomButton>
  );
};

export default Button;
