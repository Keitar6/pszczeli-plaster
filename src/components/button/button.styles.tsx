import styled from "styled-components";
import { Colors } from "../../global.styles";

export const BaseButton = styled.button`
  min-width: 10rem;
  width: auto;
  letter-spacing: 0.5px;
  // line-height: 3.1rem;
  padding: 1rem 3rem;
  font-size: 1rem;
  background-color: ${Colors.light};
  // text-transform: uppercase;
  font-family: "Open Sans Condensed";
  font-weight: bolder;
  border: none;
  cursor: pointer;
  display: flex;
  justify-content: center;
  align-items: center;
  // border: 1px solid ${Colors.primary};

  &:hover {
    background-color: ${Colors.dark};
    color: ${Colors.primary};
    // border: 1px solid black;
  }
`;

export const GoogleSignInButton = styled(BaseButton)`
  background-color: ${Colors.googleButtonBCKG};
  color: ${Colors.white};

  &:hover {
    background-color: ${Colors.googleButtonHover};
    border: none;
  }
`;

export const LoginButton = styled(BaseButton)`
  background-color: ${Colors.primary};
  color: ${Colors.dark};
  border: 1px solid none;
  border-radius: 0.6rem;
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.75));
  
  &:hover {
    background-color: ${Colors.dark};
    color: ${Colors.primary};
    border: none;
  }
`;
export const InvertedButton = styled(LoginButton)`
  background-color: ${Colors.dark};
  color: ${Colors.primary};

  &:hover {
    border: none;
    background-color: ${Colors.primary};
    color: ${Colors.dark};
  }
`;

export const UserMenuFuncButton = styled(BaseButton)`
  background-color: white;
  border: 1px solid ${Colors.dark};
  &:hover {
    background-color: ${Colors.dark};
    color: ${Colors.white};
    border: none;
  }
`;

export const ButtonSpiner = styled.div`
  display: inline-block;
  width: 30px;
  height: 30px;
  border: 3px solid rgba(195, 195, 195, 0.6);
  border-radius: 50%;
  border-top-color: #636767;
  animation: spin 1s ease-in-out infinite;
  -webkit-animation: spin 1s ease-in-out infinite;

  @keyframes spin {
    to {
      -webkit-transform: rotate(360deg);
    }
  }
  @-webkit-keyframes spin {
    to {
      -webkit-transform: rotate(360deg);
    }
  }
`;
