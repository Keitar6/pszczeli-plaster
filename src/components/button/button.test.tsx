import ComponentsRenderer from "react-test-renderer";

import Button, { BUTTON_TYPE_CLASSES, getButton } from "./button.component";
import { BaseButton } from "./button.styles";

describe("Buttons", () => {
  test("buttonType - BaseButton", () => {
    const button = ComponentsRenderer.create(
      <Button buttonType={BUTTON_TYPE_CLASSES.base} />
    ).toJSON();
    expect(button).toMatchSnapshot();
  });

  test("buttonType - productCardButton", () => {
    const button = ComponentsRenderer.create(
      <Button buttonType={BUTTON_TYPE_CLASSES.productCard}></Button>
    ).toJSON();
    expect(button).toMatchSnapshot();
  });

  test("getButton", () => {
    expect(getButton(undefined)).toEqual(BaseButton);
  });
});
