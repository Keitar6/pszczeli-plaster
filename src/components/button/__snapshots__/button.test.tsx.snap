// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons buttonType - BaseButton 1`] = `
.c0 {
  min-width: 10rem;
  width: auto;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  padding: 1rem 3rem;
  font-size: 1 rem;
  background-color: #faeff0;
  font-family: "Open Sans Condensed";
  font-weight: bolder;
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  background-color: #023047;
  color: #FFB703;
}

.c0:focus {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
  -webkit-transition: all 0.6s ease;
  transition: all 0.6s ease;
}

<button
  className="c0"
/>
`;

exports[`Buttons buttonType - productCardButton 1`] = `
.c0 {
  min-width: 10rem;
  width: auto;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  padding: 1rem 3rem;
  font-size: 1 rem;
  background-color: #faeff0;
  font-family: "Open Sans Condensed";
  font-weight: bolder;
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  background-color: #023047;
  color: #FFB703;
}

.c0:focus {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
  -webkit-transition: all 0.6s ease;
  transition: all 0.6s ease;
}

.c1 {
  background-color: #FFB703;
  color: #023047;
  border: none;
  border-radius: 0.6rem;
  -webkit-filter: drop-shadow(0px 4px 4px rgba(0,0,0,0.75));
  filter: drop-shadow(0px 4px 4px rgba(0,0,0,0.75));
}

.c1:hover {
  background-color: #023047;
  color: #FFB703;
  border: none;
}

.c2 {
  background-color: #0230473e;
  color: #FFB703;
  min-width: 5rem;
  padding: 1rem 2rem;
}

.c2:hover {
  border: none;
  background-color: #FFB7033e;
  color: #023047;
}

<button
  className="c0 c1 c2"
/>
`;
